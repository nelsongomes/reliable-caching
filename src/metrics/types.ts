export type CacheStats = {
  hits: number;
  misses: number;

  averageHitTime: number;
  averageMissTime: number;
};

export type CacheMetrics = {
  operation: string;
  // this indicates the ratio of cache hits
  hitRatio: number;
  // this indicate how much time we are saving compared with average miss time
  timeSavedRatio: number;
  // this indicates how much faster on average your request is when it hits cache
  performanceGain: number;
};
