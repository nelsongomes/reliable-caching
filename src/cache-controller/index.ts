export * from "./controller-interface";
export * from "./redis-cache-controller";
