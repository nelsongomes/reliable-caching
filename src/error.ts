import createHttpError from "http-errors";

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/no-explicit-any
export function normalizeError(error: any): Error {
  if (error instanceof Error) {
    return error;
  } else if (typeof error === "string") {
    try {
      const parsedError = JSON.parse(error);

      switch (parsedError.type) {
        case "Error":
          return Object.assign(new Error(), parsedError);
        case "EvalError":
          return Object.assign(new EvalError(), parsedError);
        case "InternalServerError":
        case "BadRequestError":
        case "NotFoundError":
          return Object.assign(
            createHttpError(
              parsedError.statusCode || 500,
              parsedError.message || "Unknown error"
            ),
            parsedError
          );
        case "RangeError":
          return Object.assign(new RangeError(), parsedError);
        case "ReferenceError":
          return Object.assign(new ReferenceError(), parsedError);
        case "SyntaxError":
          return Object.assign(new SyntaxError(), parsedError);
        case "TypeError":
          return Object.assign(new TypeError(), parsedError);
        case "URIError":
          return Object.assign(new URIError(), parsedError);
      }
    } catch (e) {
      // failed to parse JSON
      return new Error(error);
    }
  }

  return new Error(error);
}
