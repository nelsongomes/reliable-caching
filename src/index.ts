export * from "./metrics";
export * from "./keys";
export * from "./sign";
export * from "./cache-controller";
export * from "../test/utils";
