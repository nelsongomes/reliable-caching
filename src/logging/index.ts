export enum LogLevel {
  Debug = "debug",
  Error = "error",
  Info = "info",
}

export type logHandle = (
  logLevel: LogLevel,
  message: string,
  error?: unknown
) => void;
