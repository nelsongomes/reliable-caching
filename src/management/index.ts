export * from "./generic-manager";
export * from "./management-interface";
