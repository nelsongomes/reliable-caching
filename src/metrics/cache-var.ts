import { CacheStats } from "./types";

export const operationStats: Map<string, CacheStats> = new Map();
