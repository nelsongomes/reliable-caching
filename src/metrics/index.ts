export * from "./cache-stats";
export * from "./combine-stats";
