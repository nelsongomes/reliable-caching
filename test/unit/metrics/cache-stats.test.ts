import { CacheStatsManager, MAX_HISTORY } from "../../../src";
import { operationStats } from "../../../src/metrics/cache-var";
import { CacheMetrics, CacheStats } from "../../../src/metrics/types";

describe("CacheStatsManager", () => {
  it("Stats should not throw an error case there is a hit but not misses", async () => {
    const operation = "query-a";
    CacheStatsManager.hit(operation, 1);

    expect(CacheStatsManager.getOperationStats(operation)).toStrictEqual({
      operation,
      hitRatio: 1,
      performanceGain: 0,
      timeSavedRatio: 0,
    });
  });

  it("Stats should not throw an error case there is a miss but not hits", async () => {
    const operation = "query-b";
    CacheStatsManager.miss(operation, 1);

    expect(CacheStatsManager.getOperationStats(operation)).toStrictEqual({
      operation,
      hitRatio: 0,
      performanceGain: 0,
      timeSavedRatio: 0,
    });
  });

  it("Stats should not throw an error case there is no hits and misses", async () => {
    const operation = "query-c";
    CacheStatsManager.miss(operation, 1);

    expect(CacheStatsManager.getOperationStats(operation)).toStrictEqual({
      operation,
      hitRatio: 0,
      performanceGain: 0,
      timeSavedRatio: 0,
    });
  });

  it("We should avoid have too much history to avoid overflow", async () => {
    const operation = "overflow-hit";

    const currentState = {
      averageHitTime: 1,
      averageMissTime: 60000,
      hits: MAX_HISTORY + 1,
      misses: MAX_HISTORY + 1,
    };

    operationStats.clear();
    operationStats.set(operation, currentState);

    CacheStatsManager.miss(operation, 1500);
    CacheStatsManager.hit(operation, 5);

    expect(CacheStatsManager.getOperationStats(operation)).toStrictEqual({
      operation,
      hitRatio: 0.5,
      performanceGain: 59970.161935225915,
      timeSavedRatio: 0.4999916625204291,
    });
    expect(CacheStatsManager.getOperations()).toStrictEqual([operation]);
    expect(CacheStatsManager.operationStatsString(operation)).toMatchSnapshot();
    expect(CacheStatsManager.getOperationData(operation)).toStrictEqual({
      averageHitTime: 1.0004,
      averageMissTime: 59994.15,
      hits: 10001,
      misses: 10001,
    });
    expect(CacheStatsManager.operationStatsString("nonexistent")).toBe("");
  });

  it("Should match manually calculated math", async () => {
    const operation = "query-f";
    const testList: [CacheStats, CacheMetrics][] = [
      [
        {
          averageHitTime: 5,
          averageMissTime: 1500,
          hits: 1,
          misses: 1,
        },
        {
          operation,
          hitRatio: 0.5,
          performanceGain: 300,
          timeSavedRatio: 0.49833333333333335,
        },
      ],
      [
        {
          averageHitTime: 25,
          averageMissTime: 50,
          hits: 950000,
          misses: 50000,
        },
        {
          operation,
          hitRatio: 0.95,
          performanceGain: 2,
          timeSavedRatio: 0.475,
        },
      ],
      [
        {
          averageHitTime: 5,
          averageMissTime: 10,
          hits: 40,
          misses: 10,
        },
        {
          operation,
          hitRatio: 0.8,
          performanceGain: 2,
          timeSavedRatio: 0.4,
        },
      ],
      [
        {
          averageHitTime: 25,
          averageMissTime: 150,
          hits: 950000,
          misses: 50000,
        },
        {
          operation,
          hitRatio: 0.95,
          performanceGain: 6,
          timeSavedRatio: 0.7916666666666666,
        },
      ],
    ];

    testList.forEach((test) => {
      const [input, outcome] = test;

      operationStats.clear();
      operationStats.set(operation, input);

      expect(CacheStatsManager.getOperationStats(operation)).toStrictEqual(
        outcome
      );
    });
  });

  it("Should work with other sources of data", async () => {
    const operation = "query-f";

    expect(
      CacheStatsManager.operationStatsString(
        operation,
        new Map([
          [
            operation,
            { averageHitTime: 1, averageMissTime: 1, hits: 1, misses: 1 },
          ],
        ])
      )
    ).toMatchSnapshot();
  });
});
