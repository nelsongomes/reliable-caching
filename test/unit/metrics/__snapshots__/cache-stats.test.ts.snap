// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CacheStatsManager Should work with other sources of data 1`] = `
"Operation query-f Cache Metrics:
	Hit Ratio:        0.5
	Performance Gain: x1
	Processing Ratio: 0
	Time Saved Ratio: 0"
`;

exports[`CacheStatsManager We should avoid have too much history to avoid overflow 1`] = `
"Operation overflow-hit Cache Metrics:
	Hit Ratio:        0.5
	Performance Gain: x59970.161935225915
	Processing Ratio: 0.9999666506378158
	Time Saved Ratio: 0.4999916625204291"
`;
